src/main/drivers/system_stm32g4xx.c

```diff
@@ -0,0 +1,207 @@
+    //NVIC_SetVectorTable((uint32_t)&isr_vector_table_base, 0x0);
+    //__HAL_RCC_USB_OTG_FS_CLK_DISABLE;
+
+    //RCC_ClearFlag();
+
+    // Init cycle counter
+    cycleCounterInit();
+
+    // SysTick is updated whenever HAL_RCC_ClockConfig is called.
+
+    // Disable UCPD pull-down functionality on PB4
+    // AN5093 6.3.3:
+    // (after device startup this pull-down on PB4 can be disabled by
+    // setting UCPD1_DBDIS bit in the PWR_CR3 register).
+    //
+    // There is also a similar UCPD stand-by functionality control,
+    // but don't fiddle with this one. In particular, disabling it by
+    // calling HAL_PWREx_DisableUSBStandByModePD() will cause boot loader
+    // invocation to fail.
+
+#if defined(PWR_CR3_UCPD_DBDIS)
+    HAL_PWREx_DisableUSBDeadBatteryPD();
+#endif
+}
+
+void systemReset(void)
+{
+    // SCB_DisableDCache();
+    // SCB_DisableICache();
+
+    __disable_irq();
+    NVIC_SystemReset();
+}
+
+void forcedSystemResetWithoutDisablingCaches(void)
+{
+    persistentObjectWrite(PERSISTENT_OBJECT_RESET_REASON, RESET_FORCED);
+
+    __disable_irq();
+    NVIC_SystemReset();
+}
+
+void systemResetToBootloader(bootloaderRequestType_e requestType)
+{
+    switch (requestType) {
+#if defined(USE_FLASH_BOOT_LOADER)
+    case BOATLOADER_REQUEST_FLASH:
+        persistentObjectWrite(PERSISTENT_OBJECT_RESET_REASON, RESET_FLASH_BOOTLOADER_REQUEST);
+
+        break;
+#endif
+    case BOOTLOADER_REQUEST_ROM:
+    default:
+        persistentObjectWrite(PERSISTENT_OBJECT_RESET_REASON, RESET_BOOTLOADER_REQUEST_ROM);
+
+        break;
+    }
+
+    __disable_irq();
+    NVIC_SystemReset();
+}
+
+#define SYSMEMBOOT_VECTOR_TABLE ((uint32_t *)0x1fff0000)
+
+typedef void *(*bootJumpPtr)(void);
+
+void systemJumpToBootloader(void)
+{   
+    __SYSCFG_CLK_ENABLE();
+    
+    uint32_t bootStack =  SYSMEMBOOT_VECTOR_TABLE[0];
+    
+    bootJumpPtr SysMemBootJump = (bootJumpPtr)SYSMEMBOOT_VECTOR_TABLE[1];
+    
+    __set_MSP(bootStack); //Set the main stack pointer to its default values
+    
+    SysMemBootJump();
+    
+    while (1);
+}
+
+static uint32_t bootloaderRequest;
+
+void systemCheckResetReason(void)
+{
+    bootloaderRequest = persistentObjectRead(PERSISTENT_OBJECT_RESET_REASON);
+
+    switch (bootloaderRequest) {
+#if defined(USE_FLASH_BOOT_LOADER)
+    case BOATLOADER_REQUEST_FLASH:
+#endif
+    case RESET_BOOTLOADER_REQUEST_ROM:
+        persistentObjectWrite(PERSISTENT_OBJECT_RESET_REASON, RESET_BOOTLOADER_POST);
+        break;
+
+    case RESET_MSC_REQUEST:
+        // RESET_REASON will be reset by MSC
+        return;
+
+    case RESET_FORCED:
+        persistentObjectWrite(PERSISTENT_OBJECT_RESET_REASON, RESET_NONE);
+        return;
+
+    case RESET_NONE:
+        if (!(RCC->CSR & RCC_CSR_SFTRSTF)) {
+            // Direct hard reset case
+            return;
+        }
+        // Soft reset; boot loader may have been active with BOOT pin pulled high.
+        FALLTHROUGH;
+
+    case RESET_BOOTLOADER_POST:
```

Comment:
Is this two-stage reboot actually needed on this MCU type?

Reply:
Nope, and yes, it could be much much simpler ;)